aiohttp/_parser.pyx

```diff
@@ -0,0 +1,497 @@
+            payload = FlowControlStreamReader(self._protocol, loop=self._loop)
+        else:
+            payload = EMPTY_PAYLOAD
+
+        self._payload = payload
+        if encoding is not None:
+            self._payload = DeflateBuffer(payload, encoding)
+
+        self._messages.append((msg, payload))
+
+    cdef _on_message_complete(self):
+        self._payload.feed_eof()
+        self._payload = None
+
+    cdef _on_chunk_header(self):
+        if (self._current_header_value is not None or
+            self._current_header_name is not None):
+            raise BadHttpMessage('invalid headers state')
+
+        if self._proto_on_chunk_header is not None:
+            self._proto_on_chunk_header()
+
+    cdef _on_chunk_complete(self):
+        self._maybe_call_on_header()
+
+        if self._proto_on_chunk_complete is not None:
+            self._proto_on_chunk_complete()
+
+    ### Public API ###
+
+    def http_version(self):
+        cdef cparser.http_parser* parser = self._cparser
+
+        if parser.http_major == 1:
+            if parser.http_minor == 0:
+                return HttpVersion10
+            elif parser.http_minor == 1:
+                return HttpVersion11
+
+        return HttpVersion(parser.http_major, parser.http_minor)
+
+    def feed_data(self, data):
+        cdef:
+            size_t data_len
+            size_t nb
+
+        PyObject_GetBuffer(data, &self.py_buf, PyBUF_SIMPLE)
+        data_len = <size_t>self.py_buf.len
+
+        nb = cparser.http_parser_execute(
+            self._cparser,
+            self._csettings,
+            <char*>self.py_buf.buf,
+            data_len)
+
+        PyBuffer_Release(&self.py_buf)
+
+        # i am not sure about cparser.HPE_INVALID_METHOD,
+        #  seems get err for valid request
+        # test_client_functional.py::test_post_data_with_bytesio_file
+        if (self._cparser.http_errno != cparser.HPE_OK and
+                (self._cparser.http_errno != cparser.HPE_INVALID_METHOD or
+                 self._cparser.method == 0)):
+            ex = parser_error_from_errno(
+                <cparser.http_errno> self._cparser.http_errno)
+            if self._last_error is not None:
+                ex.__context__ = self._last_error
+                self._last_error = None
+            raise ex
+
+        if self._messages:
+            messages = self._messages
+            self._messages = []
+        else:
+            messages = ()
+
+        if self._upgraded:
+            return messages, True, data[nb:]
+        else:
+            return messages, False, None
+
+
+cdef class HttpRequestParser(HttpParser):
+
+    def __init__(self, protocol, loop,
+                 size_t max_line_size=8190, size_t max_headers=32768,
+                 size_t max_field_size=8190):
+         self._init(cparser.HTTP_REQUEST, protocol, loop,
+                    max_line_size, max_headers, max_field_size)
+        #self._proto_on_url = getattr(protocol, 'on_url', None)
```

Comment:
i've not decided on url, http-parser provides url parsing, should give some more performance, but it requires yarl integration